src/rust/src/declarative_asn1/decode.rs

```diff
@@ -57,6 +59,42 @@ fn decode_printable_string<'a>(
     Ok(pyo3::Bound::new(py, PrintableString { inner })?)
 }
 
+fn decode_utc_time<'a>(
+    py: pyo3::Python<'a>,
+    parser: &mut Parser<'a>,
+) -> ParseResult<pyo3::Bound<'a, UtcTime>> {
+    let value = parser.read_element::<asn1::UtcTime>()?;
+    let dt = value.as_datetime();
+
+    let inner = crate::x509::datetime_to_py_utc(py, dt)?
+        .downcast_into::<pyo3::types::PyDateTime>()?
+        .unbind();
+
+    Ok(pyo3::Bound::new(py, UtcTime { inner })?)
+}
+
+fn decode_generalized_time<'a>(
+    py: pyo3::Python<'a>,
+    parser: &mut Parser<'a>,
+) -> ParseResult<pyo3::Bound<'a, GeneralizedTime>> {
+    let value = parser.read_element::<asn1::GeneralizedTime>()?;
+    let dt = value.as_datetime();
+
+    let microseconds = match value.nanoseconds() {
+        Some(x) if x % 1_000 == 0 => Ok(x / 1_000),
+        Some(_) => Err(pyo3::exceptions::PyValueError::new_err(
+            "decoded GeneralizedTime data has higher precision than supported".to_string(),
+        )),
+        None => Ok(0),
```

Comment:
This should be `None`, instead of `Ok(0)` I think?

Reply:
Err, wait sorry I see what's going. Instead of `?` on the match block, I'd just have a `return` in the Err branch and then drop the `Ok()`

Reply:
No, we use it for `datetime_to_py_utc_with_microseconds`, which takes a `u32` for the microseconds.

Reply:
fixed